packages/apollo-server-core/src/requestPipeline.ts

```diff
@@ -284,32 +289,53 @@ export async function processGraphQLRequest<TContext>(
     // pipeline, and given plugins appropriate ability to object (by throwing
     // an error) and not actually write, we'll write to the cache if it was
     // determined earlier in the request pipeline that we should do so.
-    if (persistedQueryRegister && persistedQueryCache) {
+    if (requestContext.metrics.persistedQueryRegister && persistedQueryCache) {
       Promise.resolve(persistedQueryCache.set(queryHash, query)).catch(
         console.warn,
       );
     }
 
-    const executionDidEnd = await dispatcher.invokeDidStartHook(
-      'executionDidStart',
+    let response: GraphQLResponse | null = await dispatcher.invokeHooksUntilNonNull(
+      'responseForOperation',
       requestContext as WithRequired<
         typeof requestContext,
         'document' | 'operation' | 'operationName'
       >,
     );
+    if (response == null) {
+      const executionDidEnd = await dispatcher.invokeDidStartHook(
+        'executionDidStart',
+        requestContext as WithRequired<
+          typeof requestContext,
+          'document' | 'operation' | 'operationName'
```

Comment:
I believe `metrics` is guaranteed here?

```suggestion
          'document' | 'operation' | 'operationName' | 'metrics'
```

Also, maybe the same treatment on `didResolveOperation` above?